File: RayTracer/RayTracer/GeometricObjects/Sphere.fs
﻿namespace RayTracer.GeometricObjects

open RayTracer.BaseTypes

type Sphere(center : Vec3, radius : float) =
    interface GeometricObject with
        member this.Hit(ray : Ray) : HitRecord option=
            let temp = ray.Origin - center
            let a = (Vec3Ops.Dot ray.Direction ray.Direction)
            let b = 2.0 * (Vec3Ops.Dot temp ray.Direction)
            let c = (Vec3Ops.Dot temp temp) - radius * radius
            let disc = b * b - 4.0 * a * c

            match disc < 0.0 with
            | true -> None
            | false ->
                let t1 = (-b - sqrt(disc)) / a * 0.5;
                let t2 = (-b + sqrt(disc)) / a * 0.5;
                let t = (t1, t2)

                match t with
                | (t1, _) when t1 > kEpsilon ->
                    let position = ray.GetPoint(t1)
                    let normal = Vec3Ops.Normalize (position - center)

                    Some({ Normal = normal
                           Position = position
                           RayT = t1 })
                | (_, t2) when t2 > kEpsilon ->
                    let position = ray.GetPoint(t2)
                    let normal = Vec3Ops.Normalize (position - center)

                    Some({ Normal = normal
                           Position = position
                           RayT = t2 })
                | _ -> None

File: Source/Core/GeometricObject/Sphere.cpp
#include "Sphere.h"

void Sphere::UpdateBoundingBox()
{
    Point3D delta(Radius, Radius, Radius);
    boundingBox = BBox(Center - delta, Center + delta);
}

Sphere::Sphere(Point3D center, FP_TYPE radius) : Center(center), Radius(radius) { UpdateBoundingBox(); }

void Sphere::setCenter(Point3D center) { Center = center; UpdateBoundingBox(); }
void Sphere::setRadius(FP_TYPE radius) { Radius = radius; UpdateBoundingBox(); }
Point3D Sphere::getCenter() { return Center; }
FP_TYPE Sphere::getRadius() { return Radius; }

HitRecord Sphere::Hit(const Ray& ray)
{
    HitRecord record;
    auto temp = ray.Origin - Center;
    auto a = ray.Direction * ray.Direction;
    auto b = 2.0 * (temp * ray.Direction);
    auto c = temp * temp - Radius * Radius;
    auto disc = b * b - 4.0 * a * c;

    if (disc >= 0.0)
    {
        auto t1 = (-b - sqrt(disc)) / a * 0.5;
        auto t2 = (-b + sqrt(disc)) / a * 0.5;

        if (t1 > kEpsilon)
        {
            auto p = ray.GetPoint(t1);
            auto n = p - Center;
            record.Hit = true;
            record.Normal = (n * -ray.Direction > 0.0 ? n.normalised() : -n.normalised());
            record.HitPoint = p;
            record.MaterialPtr = materialPtr;
            record.Ray = ray;
            record.T = t1;
        }
        else if (t2 > kEpsilon)
        {
            auto p = ray.GetPoint(t2);
            auto n = p - Center;
            record.Hit = true;
            record.Normal = (n * -ray.Direction > 0.0 ? n.normalised() : -n.normalised());
            record.HitPoint = p;
            record.MaterialPtr = materialPtr;
            record.Ray = ray;
            record.T = t2;
        }
    }
    return record;
}
